mgr/dashboard: Add initiators

afreen23 opened the pull request:
## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

nizamial09 reviewed (commented):
UI changes looks good to me. I just want to make sure about the API changes and we are not making anything weirder there. I don't like the idea of abandoning the model and changing the value ourselves here. I think the best way would be to update the subsystem model with the initiators key. But I'll ask @epuertat 's help here to properly do this.

afreen23 commented:
> UI changes looks good to me. I just want to make sure about the API changes and we are not making anything weirder there. I don't like the idea of abandoning the model and changing the value ourselves here. I think the best way would be to update the subsystem model with the initiators key. But I'll ask @epuertat 's help here to properly do this.

I actually implemented first with using only subsystem model but I was getting error that grpc type `subsystems_info_cli` is not editable but read only.
Therefore I am doing the current way which is copying the fields returned from grpc response into a new response.

afreen23 commented:
recent push summary:
* fixed failing lint tests for dashboard

epuertat commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -38,12 +40,42 @@ def list(self):
-        @handle_nvmeof_error
         def list(self):
-            return NVMeoFClient().stub.list_subsystems(
+            response = []
+            # Get list of subystems
+            subsystems = NVMeoFClient().stub.list_subsystems(
                 NVMeoFClient.pb2.list_subsystems_req()
             )
+            if subsystems.status != 0:
+                raise DashboardException(
+                    msg=response.error_message,
+                    code=response.status,
+                    http_status_code=NVMeoFClient().NVMeoFError2HTTP.get(response.status, 400),
+                    component="nvmeof",
+                )
+            # Create a new response with initiators information appended
+            for subs in subsystems.subsystems:
+                res = {
+                    'nqn': subs.nqn,
+                    'enable_ha': subs.enable_ha,
+                    'serial_number': subs.serial_number,
+                    'model_number': subs.model_number,
+                    'min_cntlid': subs.min_cntlid,
+                    'max_cntlid': subs.max_cntlid,
+                    'namespace_count': subs.namespace_count,
+                    'max_namespaces': subs.max_namespaces,
+                    'subtype': subs.subtype,
+                    'initiators': [],
+                }
+                # Get list of initiators present on the subsystem
+                hosts = NVMeoFHost().list(subs.nqn)
+                # Decode the bytes object to json
+                json_str = hosts.decode('utf-8')
+                json_obj = json.loads(json_str)
+                # Appending initiators information
+                res['initiators'] = json_obj
+                response.append(res)
+            return response
```
Why changing the existing approach? It basically used decorators to programatically deal with:
* mapping gRPC entities to REST/JSON/NamedTuples,
* Error handling.

afreen23 replied:
Since this is a custom response therefore it was failing for me and I tred doing something like { subsystems: response, status: 0, status_message: "Success" } but it did not work.
If you have any suggestions I could revert then, thanks!

epuertat commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -66,13 +98,28 @@ def get(self, nqn: str):
-        def create(self, nqn: str, enable_ha: bool, max_namespaces: int = 256):
-            return NVMeoFClient().stub.create_subsystem(
+        def create(self, nqn: str, enable_ha: bool,
+                   max_namespaces: int = 256, initiators: str = None):
```
What about?
```suggestion
                   max_namespaces: int = 256, allowed_initiators: Optional[str] = None):
```

afreen23 replied:
I think its fine, to say just `initiators` , as there is nothing as non-allowed ones.

epuertat commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -66,13 +98,28 @@ def get(self, nqn: str):
                 )
             )
 
+            if subsystem_response.status != 0:
+                return subsystem_response
```
This should be done by the handle_nvmeof_error decorator.

afreen23 replied:
yes, but here we have requests followed up so checking here if all good to proceed or not.
I just copied this way from what we have been doing at rest of the places like iops api for namespaces

epuertat commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -66,13 +98,28 @@ def get(self, nqn: str):
+            if subsystem_response.status != 0:
+                return subsystem_response
+
+            if initiators is not None:
+                all_host_nqns = initiators.split(',')
+                for host_nqn in all_host_nqns:
+                    add_host_response = NVMeoFClient().stub.add_host(
+                        NVMeoFClient.pb2.add_host_req(subsystem_nqn=nqn, host_nqn=host_nqn)
+                    )
+                    if add_host_response.status != 0:
+                        return add_host_response
+
+            return subsystem_response
```
Nothing technically wrong with this but... (yeah there are BUTs :P ), as this is a "machine"/scriptable API (not directly meant for users), we wanted to keep it finegrained: "Do one single thing well". By creating a subsystem plus adding allowlisted hosts, we're breaking that rule.

We could have chosen instead the approach of having a single "God-object-like" REST Resource (Subsystems... and users could create Subsystem + listener + host allowlist + namespaces all together!). But we went the other way around: fine-grained API, mapping almost 1:1 to gRPC API.

Is there any reason to do that? From the UI it'd be just a matter of sending 2 REST requests.

afreen23 replied:
I wanted to create a UI api for this same reason, but @nizamial09 was okay so did in same api.
Yes with UI there are caveats: I could send two request from same form then need to manage when one fails other succeeds, also on deletion same sort of conflicts need to be checked. Overall it makes things on fromtend complex.

In past I tried managing multiple UI with same form and it was not a good idea from experience.
Either Ui form should be a wizard for that which is not proposed as of now.

I am okay of we create a UI API for this as that would off load most of work fro UI and keep the subsystem APi intact too

afreen23 replied:
combining this will be +1 from user experience POV. 
I think thats the advantage we want to provide users over CLI which has all of these as separate steps and commands already.

afreen23 replied:
I have revert changes, to use UI API, hence resolving.

epuertat commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -350,6 +397,73 @@ def list(self, nqn: str):
+        @empty_response
+        @handle_nvmeof_error
+        @allow_empty_body
+        def set(self, nqn: str, host_nqn: str):
+            res = NVMeoFClient().stub.list_hosts(
+                NVMeoFClient.pb2.list_hosts_req(subsystem=nqn))
+            if res.status != 0:
+                return res
+
+            new_host_nqns = host_nqn.split(',')
+            to_delete_nqns, to_add_nqns = self.determine_nqns(res, new_host_nqns)
+
+            for del_nqn in to_delete_nqns:
+                res = NVMeoFClient().stub.remove_host(
+                    NVMeoFClient.pb2.remove_host_req(subsystem_nqn=nqn, host_nqn=del_nqn)
+                )
+                if res.status != 0:
+                    return res
+                logger.info("removed host %s from subsystem %s", del_nqn, nqn)
+
+            for add_nqn in to_add_nqns:
+                res = NVMeoFClient().stub.add_host(
+                    NVMeoFClient.pb2.add_host_req(subsystem_nqn=nqn, host_nqn=add_nqn)
+                )
+                if res.status != 0:
+                    return res
+                logger.info("added host %s to subsystem %s", add_nqn, nqn)
+
+            logger.info("Updated the hosts in subsystem %s", nqn)
+            return res
```
Same question. Do we want subsystem APIs to make so many things?

afreen23 replied:
This is the basic PUT api for hosts `PUT /api/nvmeof/subsystem/{nqn}/host/{host_nqn}`

This is needed  to update hosts per subsystem. I mean if I have 10 initiators per subsystem then instead of having 10 requests I can send one request. This will be extremely helpful from UI POV.

You may check the attached screenhost of  UI where its used.

afreen23 commented:
jenkins test api

afreen23 commented:
jenkins test make check

aaSharma14 reviewed (approved):
tested locally, works fine.

afreen23 commented:
jenkins test make check arm64

afreen23 commented:
jenkins test dashboard

afreen23 commented:
Update summary:
* using UI API for all tasks , at least to have concerns separated from official APIs :)

epuertat commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -392,23 +395,148 @@ def list(self, nqn: str):
+        @Endpoint('GET', '/subsystems')
+        @ReadPermission
+        def list(self):
+            response = []
+            # Get list of subystems
+            subsystems = NVMeoFClient().stub.list_subsystems(
+                NVMeoFClient.pb2.list_subsystems_req()
+            )
+            if subsystems.status != 0:
+                raise DashboardException(
+                    msg=response.error_message,
+                    code=response.status,
+                    http_status_code=NVMeoFClient().NVMeoFError2HTTP.get(response.status, 400),
+                    component="nvmeof",
+                )
```
If you check the `handle_nvmeof_error()` decorator, it's also coping with another error condition, `grpc._channel._InactiveRpcError`. In the future we'll surely face new error conditions, while testing or from user reports. That's why my suggestion is to put error handling under a single code block (and error handler) and not spreading that throughout the codebase.

```suggestion
            # Get list of subystems
            reponse = handle_nvmeof_error(
                NVMeoFClient().stub.list_subsystems(
                    NVMeoFClient.pb2.list_subsystems_req()
                )
            )
```

epuertat commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -392,23 +395,148 @@ def list(self, nqn: str):
+            return subsystem_response
+
+        @staticmethod
+        def determine_nqns(res, new_host_nqns):
+            to_delete_nqns = []
+            to_add_nqns = []
```
This method is bit risky, as it's trying to do too much: converting an imperative interface to a declarative one. That's why we preferred to use POST and PATCH rather than PUT, and implement REST methods that mimic the CLI, rather than providing a super-high level REST API.

This REST API (nor the nvmeof CLI) doesn't provide any transaction safety guarantees, so if "someone" added/removed another host while this transaction is running (not likely, but not impossible), this method will delete/re-add that or will simply end up with a list of host NQNs different to what was stated in the REST PUT request.

afreen23 replied:
The goal is to allow user to add/delete multiple hosts rather than single one(_as done by CLI already_)
If you have any better suggestions to do that let me know, will update.

epuertat replied:
Users can still add/remove multiple hosts... Only that it'll mean more `POST`/`DELETE` requests. Personally, I still think that it's not a good thing that the Dashboard is not using its own REST API.

The current NVMe REST API allows the UI to do anything (it exposes 100% of the gRPC API), and for the scale of NVMe/TCP deployments I don't see much benefit from creating amalgamated REST endpoints (and I see the main issue that Dashboard devels will have to take care of the same gRPC-wrapping code in 2 different places).

afreen23 replied:
IMO its a good idea to have bulk operations like deleting multiple/adding multiple.
For now, both API and CLI allow user to create their own batch/bash scripts around those which is good practice as far as backend impl is concerned (like fine-grained, declarative and maintainable).

But for UI,if we do same then its no value and we are not providing anything valuable over what CLI has to do.
At least, I wont use the UI then over CLI if its same.

Though its easy for me as dev to maintain such UI and get the task done yet I dont see a benefit for dashboard user.

epuertat commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -392,23 +395,148 @@ def list(self, nqn: str):
+            current_host_nqns = [host.nqn for host in res.hosts]
+
+            # All hosts are allowed but request is to add new hosts
+            if res.allow_any_host and new_host_nqns != ["*"]:
+                to_delete_nqns = ["*"]
+                to_add_nqns = new_host_nqns
+
+            # All hosts are not allowed but request is to allow all
+            elif new_host_nqns == ["*"]:
+                to_delete_nqns = current_host_nqns
+                to_add_nqns.append("*")
```
Why? Even if it doesn't (apparently) make much sense, the CLI supports adding both `*` and specific hosts NQNs. It's common with access rules that you can add contradictory rules like the following:
```
allow all
allow user X do that
```
The main reason for that is that in some situations you just want to temporarily grant extra permissions (troubleshooting, etc.) without deleting the existing rulesets.

afreen23 replied:
If both are allowed then lets do the same from UI too, makes sense 👍🏿

epuertat commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -392,23 +395,148 @@ def list(self, nqn: str):
+                # Get list of initiators present on the subsystem
+                hosts = NVMeoFHost().list(subs.nqn)
+                # Decode the bytes object to json
+                json_str = hosts.decode('utf-8')
+                json_obj = json.loads(json_str)
```
What type of Python object is this? A list? In that case, instead of converting to/from JSON, you can probably do something like:
```suggestion
                hosts = [host for host in hosts]
```

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.ts:
```diff
@@ -51,6 +55,24 @@ export class NvmeofSubsystemsFormComponent implements OnInit {
+        this.nvmeofService
+          .getInitiators(params.subsystem_nqn)
+          .subscribe((initiators: NvmeofSubsystemHost[]) => {
+            if (initiators[0].nqn !== '*') {
+              this.subsystemForm.get('initiators').setValue('add-host');
+              initiators.forEach((init: NvmeofSubsystemHost) => this.addHost(init.nqn));
+            }
```
Even if it looks like a smart UX move, as the CLI supports specifying both `*` and specific NQNs, we shouldn't rule out deployments where both things are set... And this makes impossible adding specific NQNs if `*` is already set.

afreen23 replied:
If both are allowed then lets do the same from UI too, makes sense 👍🏿

epuertat reviewed (commented):
I still think that:
* This is trying to do too many things (e.g.: converting an imperative API to a declarative one),
* it is creating an "unofficial" shortcut to the REST API (as a result of it, the official subsystem API will no longer be used then),
* IMHO the existing API was enough.

nizamial09 commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -392,23 +395,148 @@ def list(self, nqn: str):
+        @ReadPermission
+        @EndpointDoc("Display NVMe/TCP service Status",
+                     responses={200: NVME_SCHEMA})
+        def status(self) -> dict:
+            status: Dict[str, Any] = {'available': True, 'message': None}
+            orch_backend = mgr.get_module_option_ex('orchestrator', 'orchestrator')
+            if orch_backend == 'cephadm':
+                orch = OrchClient.instance()
+                orch_status = orch.status()
+                if not orch_status['available']:
+                    return status
+                if not orch.services.list_daemons(daemon_type='nvmeof'):
+                    status["available"] = False
+                    status["message"] = 'Create an NVMe/TCP service to get started.'
+            return status
```
@epuertat: i think `status` was introduced as a `ui-api` router by @afreen23 recently and no, it is not used in the plug-in.

epuertat replied:
thanks @nizamial09 !

anthonyeleven commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -392,23 +395,148 @@ def list(self, nqn: str):
+    class NVMeoFTcpUI(BaseController):
+        @Endpoint('GET', '/status')
+        @ReadPermission
+        @EndpointDoc("Display NVMe/TCP service Status",
```
`status` lowercase?

anthonyeleven commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -392,23 +395,148 @@ def list(self, nqn: str):
+                    return status
+                if not orch.services.list_daemons(daemon_type='nvmeof'):
+                    status["available"] = False
+                    status["message"] = 'Create an NVMe/TCP service to get started.'
```
`An NVMe/TCP service must be created before it may be listed`

anthonyeleven commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -392,23 +395,148 @@ def list(self, nqn: str):
-                status["available"] = False
-                status["message"] = 'Create an NVMe/TCP service to get started.'
-        return status
+            logger.info("Updated the hosts in subsystem %s", nqn)
```
s/in/for/

anthonyeleven commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-subsystems-form/nvmeof-subsystems-form.component.html:
```diff
@@ -61,6 +61,67 @@
+                        i18n>Expected NQN format<br/>&lt;<code>nqn.$year-$month.$reverseDomainName:$utf8-string</code>".&gt; or <br/>&lt;<code>nqn.2014-08.org.nvmexpress:uuid:$UUID-string</code>".&gt;</span>
+                  <span class="invalid-feedback"
+                        *ngIf="subsystemForm.get('hosts').controls[hi].errors?.maxLength"
+                        i18n>An NQN should not be more than 223 bytes in length.</span>
```
s/should/may/

afreen23 commented:
Update Summary:

* two UI-apis for create and deleting multiple hosts
* based on review-comment and offline discussion added a warning for "Allow any host" case since both * or hosts can coexist
* updated screenshot
* addressed review-comments

anthonyeleven commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -392,23 +393,65 @@ def list(self, nqn: str):
+        @EndpointDoc("Add one or more initiator hosts to a subsystem",
+                     parameters={
+                         'subsystem_nqn': (str, 'Subsystem NQN'),
+                         "host_nqn": Param(str, 'Comma separated list of NVMeoF host NQN.'),
```
NQNs plural, and take out the period

anthonyeleven commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -392,23 +393,65 @@ def list(self, nqn: str):
+            return status
+
+        @Endpoint('POST', "/subsystem/{subsystem_nqn}/host")
+        @EndpointDoc("Add one or more initiator hosts to a subsystem",
```
To an NVMeoF subsystem?

afreen23 commented:
Update Summary:

* addressed reviews
* cleanup of some nits

anthonyeleven commented:
Docs lgtm but I’m not qualified to bless the code

afreen23 commented:
jenkins test api

cloudbehl commented:
Some recommendation for workflow

Quick one, can we have `add host`  pre-checked.

Also, Allow any host can be sub part of it.

- If Allow any host  is checked,  then we can disable adding single host/initiators.
- if user clicks create second time when we already have Allow any host, then they needs to uncheck Allow any host add single initiator. Otherwise if he doesn't uncheck then they can't do anything.

afreen23 commented:
> I still think that:
> 
>     * This is trying to do too many things (e.g.: converting an imperative API to a declarative one),
> 
>     * it is creating an "unofficial" shortcut to the REST API (as a result of it, the official subsystem API will no longer be used then),
> 
>     * IMHO the existing API was enough.

Updated the code now to cover these comments only. 
Almost re-wrote entire thing.
The existing API are very thin and fine-grained for specific  task which is nice.
But on UI I have to allow adding more than one initiator therefore just added two new UI APIs.

https://github.com/ceph/ceph/pull/58534#discussion_r1691350492

Thanks

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof-initiators-form/nvmeof-initiators-form.component.html:
```diff
@@ -0,0 +1,105 @@
+<cd-modal [pageURL]="pageURL"
+          [modalRef]="activeModal">
```
don't need this.
```suggestion
```

nizamial09 reviewed (approved):
I am okay with the UIRouter for doing the multiple-addition/deletion for now. I think its better than looping in on the UI. Thank you for taking care of this PR. Tested, LGTM

nizamial09 commented:
merging it. we can take the API improvements if any in later PRs
